sys/src/ape/cmd/pax/pax.1
.\" $Id: pax.1,v 1.2 89/02/12 10:08:47 mark Exp $
.TH PAX 1 "USENIX Association" ""
.SH NAME
pax \- portable archive exchange 
.SH SYNOPSIS
.TP \w'\fBpax\fR\ 'u
.B pax
.RB [ \-cimopuvy ]
.RI "[\fB\-f\fR" " archive" ]
.RI "[\fB\-s\fR" " replstr" ]
.RI "[\fB\-t\fR" " device" ] 
.RI [ pattern... ]
.TP \w'\fBpax\ \-r\fR\ 'u
.B "pax\ \-r"
.RB [ \-cimnopuvy ]
.RI "[\fB\-f\fR" " archive" ]
.RI "[\fB\-s\fR" " replstr" ] 
.RI "[\fB\-t\fR" " device" ] 
.RI [ pattern... ]
.TP \w'\fBpax\ \-w\fR\ 'u
.B "pax\ \-w"
.RB [ \-adimuvy ] 
.RI "[\fB\-b\fR" " blocking" ]
.RI "[\fB\-f\fR" " archive" ] 
.RI "[\fB\-s\fR" " replstr" ] 
.RI "[\fB\-t\fR" " device" ] 
.RI "[\fB\-x\fR" " format" ]
.RI [ pathname... ]
.TP \w'\fBpax\ \-rw\fR\ 'u
.B "pax\ \-rw"
.RB [ \-ilmopuvy ] 
.RI "[\fB\-s\fR" " replstr" ] 
.RI [ pathname... ]
directory
.SH DESCRIPTION
.I Pax 
reads and writes archive files which conform to the 
.B "Archive/Interchange File Format" 
specified in 
.IR "IEEE Std. 1003.1-1988" .
.I Pax
can also read, but not write, a number of other file formats
in addition to those specified in the 
.B "Archive/Interchange File Format"
description.
Support for these traditional file formats, such as V7 
.I "tar" 
and System V binary 
.I "cpio" 
format archives,
is provided for backward compatibility and to maximize portability.
.PP
.I Pax 
will also support traditional 
.I cpio 
and 
System V
.I tar
interfaces if invoked with the name
"cpio" or "tar" respectively.
See the 
.I cpio(1) 
or 
.I tar(1)
manual pages for more details.
.PP
Combinations of the
.B \-r 
and 
.B \-w
command line arguments specify whether 
.I pax 
will read, write or list the contents of the specified archive,
or move the specified files to another directory.
.PP
The command line arguments are:
.TP .5i
.B \-w
writes the files and directories specified by 
.I pathname
operands to the standard output together with the pathname and status 
information prescribed by the archive format used.
A directory
.I pathname
operand refers to the files and (recursively) subdirectories of that 
directory.
If no 
.I pathname 
operands are given, then the standard input is read to get a
list of pathnames to copy, one pathname per line.
In this case, only those pathnames appearing on the standard input are
copied.
.TP .5i
.B \-r
.I Pax
reads an archive file from the standard input.
Only files with names that match any of the 
.I pattern
operands are selected for extraction.
The selected files are conditionally created and copied relative to the
current directory tree, subject to the options described below.
By default, the owner and group of selected files will be that of the
invoking process, and the permissions and modification times will be the
sames as those in the archive.
.RS .5i
.PP
The supported archive formats are automatically detected on input.
The default output format is 
.IR ustar ,
but may be overridden by the 
.B \-x
.I format
option described below.
.RE
.TP .5i
.B \-rw
.I Pax 
reads the files and directories named in the
.I pathname
operands and copies them to the destination 
.IR directory .
A directory
.I pathname 
operand refers to the files and (recursively) subdirectories of that
directory.
If no 
.I pathname
operands are given, the standard input is read to get a list of pathnames
to copy, one pathname per line.
In this case, only those pathnames appearing on the standard input are
copied.
The directory named by the 
.I directory
operand must exist and have the proper permissions before the copy can
occur.
.PP
If neither the
.BR \-r " or " \-w 
options are given, then 
.I pax
will list the contents of the specified archive.
In this mode, 
.I pax 
lists normal files one per line, hard link pathnames as
.sp
.RS 1i
.IR pathname " == " linkname
.RE
.sp
and symbolic link pathnames (if supported by the implementation) as
.sp
.RS 1i
.IR pathname " -> " linkname
.RE
.sp
where 
.I pathname
is the name of the file being extracted, and
.I linkname
is the name of a file which appeared earlier in the archive.
.PP
If the 
.B \-v
option is specified, then 
.I pax 
list normal pathnames in the same format used by the 
.I ls
utility with the
.B \-l
option.
Hard links are shown as
.sp
.RS 1i
.IR "<ls -l listing>" " == " linkname
.RE
.sp
and symbolic links (if supported) are shown as
.sp
.RS 1i
.IR "<ls -l listing>" " -> " linkname
.RE
.sp 
.PP
.I Pax 
is capable of reading and writing archives which span multiple physical 
volumes.
Upon detecting an end of medium on an archive which is not yet completed,
.I pax 
will prompt the user for the next volume of the archive and will allow the 
user to specify the location of the next volume.
.SS Options
The following options are available:
.TP 1i
.B \-a
The files specified by
.I pathname
are appended to the specified archive.
.TP 1i
.BI \-b " blocking"
Block the output at
.I blocking
bytes per write to the archive file.
A 
.B k
suffix multiplies 
.I blocking 
by 1024, a
.B b 
suffix multiplies 
.I blocking 
by 512 and a 
.B m
suffix multiplies 
.I blocking 
by 1048576 (1 megabyte).
For machines with 16-bit int's (VAXen, XENIX-286, etc.), 
the maximum buffer size is 32k-1.
If not specified,
.I blocking 
is automatically determined on input and is ignored for 
.B \-rw.
.TP 1i
.B \-c
Complement the match sense of the 
.I pattern
operands.
.TP 1i
.B \-d
Intermediate directories not explicitly listed in the archive are not 
created.
This option is ignored unless
the 
.B \-r
option is specified.
.TP 1i
.BI \-f " archive"
The
.I archive
option specifies the pathname of the input or output archive,
overriding the default of standard input for 
.B \-r
or standard output for
.BR \-w .
.TP 1i
.B \-i
Interactively rename files.
Substitutions specified by 
.B \-s
options (described below)
are performed before requesting the new file name from the user.
A file is skipped if an empty line is entered and
.I pax
exits with an exit status of 0 if 
.B EOF 
is encountered.
.TP 1i
.B \-l
Files are linked rather than copied when possible.
.TP 1i
.B \-m
File modification times are not retained.
.TP 1i
.B \-n
When 
.B \-r
is specified, but
.B \-w
is not, the 
.I pattern
arguments are treated as ordinary file names.
Only the first occurrence of each of these files in the input archive
is read.
The
.B pax
utility exits with a zero exit status after all files in the list have been
read.
If one or more files in the list is not found,
.B pax
writes a diagnostic to standard error for each of the files and exits with
a non-zero exit status.
the file names are compared before any of the 
.BR \-i ,
.BR \-s ,
or
.B \-y
options are applied.
.TP 1i
.B \-o
Restore file ownership as specified in the archive.
The invoking process must have appropriate privileges to accomplish this.
.TP 1i
.B \-p
Preserve the access time of the input files after they have been copied.
.TP 1i
.BI \-s " replstr"
File names are modified according to the substitution expression using the 
syntax of 
.I "ed(1)"
as shown:
.sp
.RS 2i
-s /\fIold\fR/\fInew\fR/\fB[\fRgp\fB]\fR
.RE
.RS 1i
.PP
Any non null character may be used as a delimiter (a / is used here as an
example).
Multiple 
.B \-s
expressions may be specified; the expressions are applied in the order
specified terminating with the first successful substitution.
The optional trailing 
.B p
causes successful mappings to be listed on standard error.
The optional trailing
.B g
causes the 
.I old 
expression to be replaced each time it occurs in the source string.
Files that substitute to an empty string are ignored both on input and
output.
.RE
.TP 1i
.BI \-t " device"
The
.I device
option argument is an implementation-defined identifier that names the input
or output archive device, overriding the default of standard input for
.B \-r
and standard output for
.BR \-w .
.TP 1i
.B \-u
Copy each file only if it is newer than a pre-existing file with the same
name.
This implies 
.BR \-a .
.TP 1i
.B \-v
List file names as they are encountered.
Produces a verbose table of contents listing on the standard output when both 
.B \-r 
and
.B \-w
are omitted,
otherwise the file names are printed to standard error as they are
encountered in the archive.
.TP 1i
.BI \-x " format"
Specifies the output archive
.IR format .
The input format, which must be one of the following, is automatically
determined when the
.B \-r 
option is used.
The supported formats are:
.RS 1i
.TP 0.75i
.I cpio
The extended 
.I CPIO
interchange format specified in
.B "Extended CPIO Format" in 
.I "IEEE Std. 1003.1-1988."
.TP 0.75i
.I ustar
The extended 
.I TAR
interchange format specified in
.B "Extended TAR Format" in 
.I "IEEE Std. 1003.1-1988."
This is the default archive format.
.RE
.TP 1i
.B \-y
Interactively prompt for the disposition of each file.
Substitutions specified by 
.B \-s
options (described above)
are performed before prompting the user for disposition.
.B EOF
or an input line starting with the character
.B q
caused
.I pax
to exit.
Otherwise, an input line starting with anything other than 
.B y
causes the file to be ignored.
This option cannot be used in conjunction with the 
.B \-i 
option.
.PP
Only the last of multiple 
.B \-f 
or
.B \-t
options take effect.
.PP
When writing to an archive, the
standard input is used as a list of pathnames if no
.I pathname
operands are specified.
The format is one pathname per line.
Otherwise, the standard input is the archive file,
which is formatted according to one of the specifications in
.B "Archive/Interchange File format"
in 
.IR "IEEE Std. 1003.1-1988" ,
or some other implementation-defined format.
.PP
The user ID and group ID of the process, together with the appropriate
privileges, affect the ability of 
.I pax
to restore ownership and permissions attributes of the archived files.
(See 
.I "format-reading utility"
in
.B "Archive/Interchange File Format"
in 
.IR "IEEE Std. 1003.1-1988" ".)"
.PP
The options
.BR \-a ,
.BR \-c ,
.BR \-d ,
.BR \-i ,
.BR \-l ,
.BR \-p ,
.BR \-t ,
.BR \-u ,
and
.BR \-y 
are provided for functional compatibility with the historical
.I cpio
and
.I tar
utilities.
The option defaults were chosen based on the most common usage of these
options, therefore, some of the options have meanings different than
those of the historical commands.
.SS Operands
The following operands are available:
.TP 1i
.I directory
The destination directory pathname for copies when both the
.B \-r
and 
.B \-w
options are specified.
The directory must exist and be writable before the copy or and error
results.
.TP 1i
.I pathname
A file whose contents are used instead of the files named on the standard
input.
When a directory is named, all of its files and (recursively) 
subdirectories
are copied as well.
.TP 1i
.IR pattern 
A
.I pattern
is given in the standard shell pattern matching notation.
The default if no 
.I pattern 
is  specified is
.BR * \,
which selects all files.
.SH EXAMPLES
The following command
.sp
.RS 1i
pax \-w \-f /dev/rmt0 \.
.RE
.sp
copies the contents of the current directory to tape drive 0.
.PP
The commands
.sp
.RS 1i
.RI mkdir " newdir"
.br
.RI cd " olddir"
.br
.RI "pax -rw . " newdir
.RE
.sp
copies the contents of 
.I olddir 
to 
.I newdir .
.PP
The command
.sp
.RS 1i
pax \-r \-s ',//*usr//*,,' -f pax.out
.RE
.sp
reads the archive 
.B pax.out
with all files rooted in "/usr" in the archive extracted
relative to the current directory.
.SH FILES
.TP 1i
/dev/tty
used to prompt the user for information when the
.BR \-i " or " \-y
options are specified. 
.SH "SEE ALSO"
cpio(1), find(1), tar(1), cpio(5), tar(5)
.SH DIAGNOSTICS
.I Pax
will terminate immediately, without processing any 
additional files on the command line or in the archive.
.SH "EXIT CODES"
.I Pax 
will exit with one of the following values:
.IP 0 .5i
All files in the archive were processed successfully.
.IP ">0" .5i
.I Pax 
aborted due to errors encountered during operation.
.SH BUGS
Special permissions may be required to copy or extract special files.
.PP
Device, user ID, and group ID numbers larger than 65535 cause additional
header records to be output.
These records are ignored by some historical version of
.I "cpio(1)"
and
.IR "tar(1)" .
.PP
The archive formats described in
.B "Archive/Interchange File Format" 
have certain restrictions that have
been carried over from historical usage.
For example, there are restrictions on the length of pathnames stored in
the archive.
.PP
When getting an "ls -l" style listing on
.I tar
format archives, link counts are listed as zero since the 
.I ustar
archive format does not keep link count information.
.PP
On 16 bit architectures, the largest buffer size is 32k-1.
This is due, in part, to using integers in the buffer allocation schemes,
however, on many of these machines, it is not possible to allocate blocks
of memory larger than 32k.
.SH COPYRIGHT
Copyright (c) 1989 Mark H. Colburn.  
.br
All rights reserved.
.PP
Redistribution and use in source and binary forms are permitted
provided that the above copyright notice is duplicated in all such 
forms and that any documentation, advertising materials, and other 
materials related to such distribution and use acknowledge that the 
software was developed by Mark H. Colburn and sponsored by The 
USENIX Association. 
.PP
THE SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
.SH AUTHOR
Mark H. Colburn
.br
Minnetech Consulting, Inc.
.br
117 Mackubin Street, Suite 1
.br
St. Paul, MN 55102
.br
mark@jhereg.MN.ORG
.sp 2
Sponsored by
.B "The USENIX Association"
for public distribution.


sys/src/ape/cmd/pax/pax.c
/*
 * This file is part of the UCB release of Plan 9. It is subject to the license
 * terms in the LICENSE file found in the top-level directory of this
 * distribution and at http://akaros.cs.berkeley.edu/files/Plan9License. No
 * part of the UCB release of Plan 9, including this file, may be copied,
 * modified, propagated, or distributed except according to the terms contained
 * in the LICENSE file.
 */

/* $Source: /u/mark/src/pax/RCS/pax.c,v $
 *
 * $Revision: 1.2 $
 *
 * DESCRIPTION
 *
 *	Pax is the archiver described in IEEE P1003.2.  It is an archiver
 *	which understands both tar and cpio archives and has a new interface.
 *
 * SYNOPSIS
 *
 *	pax -[cimopuvy] [-f archive] [-s replstr] [-t device] [pattern...]
 *	pax -r [-cimopuvy] [-f archive] [-s replstr] [-t device] [pattern...]
 *	pax -w [-adimuvy] [-b blocking] [-f archive] [-s replstr]...]
 *	       [-t device][-x format][pathname...]
 *	pax -r -w [-ilmopuvy][-s replstr][pathname...] directory
 *
 * DESCRIPTION
 *
 * 	PAX - POSIX conforming tar and cpio archive handler.  This
 *	program implements POSIX conformant versions of tar, cpio and pax
 *	archive handlers for UNIX.  These handlers have defined befined
 *	by the IEEE P1003.2 commitee.
 *
 * COMPILATION
 *
 *	A number of different compile time configuration options are
 *	available, please see the Makefile and config.h for more details.
 *
 * AUTHOR
 *
 *     Mark H. Colburn, NAPS International (mark@jhereg.mn.org)
 *
 *
 * Sponsored by The USENIX Association for public distribution.
 *
 * Copyright (c) 1989 Mark H. Colburn.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms are permitted
 * provided that the above copyright notice is duplicated in all such
 * forms and that any documentation, advertising materials, and other
 * materials related to such distribution and use acknowledge that the
 * software was developed * by Mark H. Colburn and sponsored by The
 * USENIX Association.
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 *
 * $Log:	pax.c,v $
 * Revision 1.2  89/02/12  10:05:17  mark
 * 1.2 release fixes
 *
 * Revision 1.1  88/12/23  18:02:23  mark
 * Initial revision
 *
 */

#ifndef lint
static char* ident = "$Id: pax.c,v 1.2 89/02/12 10:05:17 mark Exp $";
static char* copyright =
    "Copyright (c) 1989 Mark H. Colburn.\nAll rights reserved.\n";
#endif /* ! lint */

/* Headers */

#define NO_EXTERN
#include "pax.h"

/* Globally Available Identifiers */

char* ar_file;                     /* File containing name of archive */
char* bufend;                      /* End of data within archive buffer */
char* bufstart;                    /* Archive buffer */
char* bufidx;                      /* Archive buffer index */
char* myname;                      /* name of executable (argv[0]) */
char** n_argv;                     /* Argv used by name routines */
int n_argc;                        /* Argc used by name routines */
int archivefd;                     /* Archive file descriptor */
int blocking;                      /* Size of each block, in records */
int gid;                           /* Group ID */
int head_standard;                 /* true if archive is POSIX format */
int ar_interface;                  /* defines interface we are using */
int ar_format;                     /* defines current archve format */
int mask;                          /* File creation mask */
int ttyf;                          /* For interactive queries */
int uid;                           /* User ID */
int names_from_stdin;              /* names for files are from stdin */
OFFSET total;                      /* Total number of bytes transferred */
int16_t f_access_time;             /* Reset access times of input files */
int16_t areof;                     /* End of input volume reached */
int16_t f_dir_create;              /* Create missing directories */
int16_t f_append;                  /* Add named files to end of archive */
int16_t f_create;                  /* create a new archive */
int16_t f_extract;                 /* Extract named files from archive */
int16_t f_follow_links;            /* follow symbolic links */
int16_t f_interactive;             /* Interactivly extract files */
int16_t f_linksleft;               /* Report on unresolved links */
int16_t f_list;                    /* List files on the archive */
int16_t f_modified;                /* Don't restore modification times */
int16_t f_verbose;                 /* Turn on verbose mode */
int16_t f_link;                    /* link files where possible */
int16_t f_owner;                   /* extract files as the user */
int16_t f_pass;                    /* pass files between directories */
int16_t f_newer;                   /* append files to archive if newer */
int16_t f_disposition;             /* ask for file disposition */
int16_t f_reverse_match;           /* Reverse sense of pattern match */
int16_t f_mtime;                   /* Retain file modification time */
int16_t f_unconditional;           /* Copy unconditionally */
time_t now = 0;                    /* Current time */
uint arvolume;                     /* Volume number */
uint blocksize = BLOCKSIZE;        /* Archive block size */
FILE* msgfile;                     /* message outpu file stdout/stderr */
Replstr* rplhead = (Replstr*)NULL; /*  head of replstr list */
Replstr* rpltail;                  /* pointer to tail of replstr list */

/* Function Prototypes */

#ifdef __STDC__

static void usage(void);
static OFFSET pax_optsize(char*);

#else /* !__STDC__ */

static void usage();
static OFFSET pax_optsize();

#endif /* __STDC__ */

/* main - main routine for handling all archive formats.
 *
 * DESCRIPTION
 *
 * 	Set up globals and call the proper interface as specified by the user.
 *
 * PARAMETERS
 *
 *	int argc	- count of user supplied arguments
 *	char **argv	- user supplied arguments
 *
 * RETURNS
 *
 *	Returns an exit code of 0 to the parent process.
 */

#ifdef __STDC__

int
main(int argc, char** argv)

#else

int main(argc, argv) int argc;
char** argv;

#endif
{
	/* strip the pathname off of the name of the executable */
	if((myname = strrchr(argv[0], '/')) != (char*)NULL) {
		myname++;
	} else {
		myname = argv[0];
	}

	/* set upt for collecting other command line arguments */
	name_init(argc, argv);

	/* get all our necessary information */
	mask = umask(0);
	uid = getuid();
	gid = getgid();
	now = time((time_t*)0);

	/* open terminal for interactive queries */
	ttyf = open_tty();

	if(strcmp(myname, "tar") == 0) {
		do_tar(argc, argv);
	} else if(strcmp(myname, "cpio") == 0) {
		do_cpio(argc, argv);
	} else {
		do_pax(argc, argv);
	}
	exit(0);
	/* NOTREACHED */
}

/* do_pax - provide a PAX conformant user interface for archive handling
 *
 * DESCRIPTION
 *
 *	Process the command line parameters given, doing some minimal sanity
 *	checking, and then launch the specified archiving functions.
 *
 * PARAMETERS
 *
 *    int ac		- A count of arguments in av.  Should be passed argc
 *			  from main
 *    char **av		- A pointer to an argument list.  Should be passed
 *			  argv from main
 *
 * RETURNS
 *
 *    Normally returns 0.  If an error occurs, -1 is returned
 *    and state is set to reflect the error.
 *
 */

#ifdef __STDC__

int
do_pax(int ac, char** av)

#else

int do_pax(ac, av) int ac; /* argument counter */
char** av;                 /* arguments */

#endif
{
	int c;
	char* dirname;
	Stat st;

	/* default input/output file for PAX is STDIN/STDOUT */
	ar_file = "-";

	/*
	 * set up the flags to reflect the default pax inteface.  Unfortunately
	 * the pax interface has several options which are completely opposite
	 * of the tar and/or cpio interfaces...
	 */
	f_unconditional = 1;
	f_mtime = 1;
	f_dir_create = 1;
	f_list = 1;
	blocksize = 0;
	blocking = 0;
	ar_interface = PAX;
	ar_format = TAR; /* default interface if none given for -w */
	msgfile = stdout;

	while((c = getopt(ac, av, "ab:cdf:ilmoprs:t:uvwx:y")) != EOF) {
		switch(c) {
		case 'a':
			f_append = 1;
			f_list = 0;
			break;
		case 'b':
			if((blocksize = pax_optsize(optarg)) == 0) {
				fatal("Bad block size");
			}
			break;
		case 'c':
			f_reverse_match = 1;
			break;
		case 'd':
			f_dir_create = 0;
			break;
		case 'f':
			if(blocksize == 0) {
				blocking = 1;
				blocksize = 1 * BLOCKSIZE;
			}
			ar_file = optarg;
			break;
		case 'i':
			f_interactive = 1;
			break;
		case 'l':
			f_link = 1;
			break;
		case 'm':
			f_mtime = 0;
			break;
		case 'o':
			f_owner = 1;
			break;
		case 'p':
			f_access_time = 1;
			break;
		case 'r':
			if(f_create) {
				f_create = 0;
				f_pass = 1;
			} else {
				f_list = 0;
				f_extract = 1;
			}
			msgfile = stderr;
			break;
		case 's':
			add_replstr(optarg);
			break;
		case 't':
			if(blocksize == 0) {
				blocking = 1;
				blocksize = 10 * BLOCKSIZE;
			}
			ar_file = optarg;
			break;
		case 'u':
			f_unconditional = 1;
			break;
		case 'v':
			f_verbose = 1;
			break;
		case 'w':
			if(f_extract) {
				f_extract = 0;
				f_pass = 1;
			} else {
				f_list = 0;
				f_create = 1;
			}
			msgfile = stderr;
			break;
		case 'x':
			if(strcmp(optarg, "ustar") == 0) {
				ar_format = TAR;
			} else if(strcmp(optarg, "cpio") == 0) {
				ar_format = CPIO;
			} else {
				usage();
			}
			break;
		case 'y':
			f_disposition = 1;
			break;
		default:
			usage();
		}
	}

	if(blocksize == 0) {
		blocking = 1;
		blocksize = blocking * BLOCKSIZE;
	}
	buf_allocate((OFFSET)blocksize);

	if(f_extract || f_list) {
		open_archive(AR_READ);
		get_archive_type();
		read_archive();
	} else if(f_create) {
		if(optind >= n_argc) {
			names_from_stdin++; /* args from stdin */
		}
		open_archive(AR_WRITE);
		create_archive();
	} else if(f_append) {
		open_archive(AR_APPEND);
		get_archive_type();
		append_archive();
	} else if(f_pass && optind < n_argc) {
		dirname = n_argv[--n_argc];
		if(LSTAT(dirname, &st) < 0) {
			fatal(strerror());
		}
		if((st.sb_mode & S_IFMT) != S_IFDIR) {
			fatal("Not a directory");
		}
		if(optind >= n_argc) {
			names_from_stdin++; /* args from stdin */
		}
		pass(dirname);
	} else {
		usage();
	}

	return (0);
}

/* get_archive_type - determine input archive type from archive header
 *
 * DESCRIPTION
 *
 * 	reads the first block of the archive and determines the archive
 *	type from the data.  If the archive type cannot be determined,
 *	processing stops, and a 1 is returned to the caller.  If verbose
 *	mode is on, then the archive type will be printed on the standard
 *	error device as it is determined.
 *
 * FIXME
 *
 *	be able to understand TAR and CPIO magic numbers
 */

#ifdef __STDC__

void
get_archive_type(void)

#else

void
get_archive_type()

#endif
{
	if(ar_read() != 0) {
		fatal("Unable to determine archive type.");
	}
	if(strncmp(bufstart, "070707", 6) == 0) {
		ar_format = CPIO;
		if(f_verbose) {
			fputs("CPIO format archive\n", stderr);
		}
	} else if(strncmp(&bufstart[257], "ustar", 5) == 0) {
		ar_format = TAR;
		if(f_verbose) {
			fputs("USTAR format archive\n", stderr);
		}
	} else {
		ar_format = TAR;
	}
}

/* pax_optsize - interpret a size argument
 *
 * DESCRIPTION
 *
 * 	Recognizes suffixes for blocks (512-bytes), k-bytes and megabytes.
 * 	Also handles simple expressions containing '+' for addition.
 *
 * PARAMETERS
 *
 *    char 	*str	- A pointer to the string to interpret
 *
 * RETURNS
 *
 *    Normally returns the value represented by the expression in the
 *    the string.
 *
 * ERRORS
 *
 *	If the string cannot be interpretted, the program will fail, since
 *	the buffering will be incorrect.
 *
 */

#ifdef __STDC__

static OFFSET
pax_optsize(char* str)

#else

static OFFSET pax_optsize(str) char* str; /* pointer to string to interpret */

#endif
{
	char* idx;
	OFFSET number; /* temporary storage for current number */
	OFFSET result; /* cumulative total to be returned to caller */

	result = 0;
	idx = str;
	for(;;) {
		number = 0;
		while(*idx >= '0' && *idx <= '9')
			number = number * 10 + *idx++ - '0';
		switch(*idx++) {
		case 'b':
			result += number * 512L;
			continue;
		case 'k':
			result += number * 1024L;
			continue;
		case 'm':
			result += number * 1024L * 1024L;
			continue;
		case '+':
			result += number;
			continue;
		case '\0':
			result += number;
			break;
		default:
			break;
		}
		break;
	}
	if(*--idx) {
		fatal("Unrecognizable value");
	}
	return (result);
}

/* usage - print a helpful message and exit
 *
 * DESCRIPTION
 *
 *	Usage prints out the usage message for the PAX interface and then
 *	exits with a non-zero termination status.  This is used when a user
 *	has provided non-existant or incompatible command line arguments.
 *
 * RETURNS
 *
 *	Returns an exit status of 1 to the parent process.
 *
 */

#ifdef __STDC__

static void
usage(void)

#else

static void
usage()

#endif
{
	fprintf(stderr, "Usage: %s -[cimopuvy] [-f archive] [-s replstr] [-t "
	                "device] [pattern...]\n",
	        myname);
	fprintf(stderr, "       %s -r [-cimopuvy] [-f archive] [-s replstr] "
	                "[-t device] [pattern...]\n",
	        myname);
	fprintf(stderr, "       %s -w [-adimuvy] [-b blocking] [-f archive] "
	                "[-s replstr]\n              [-t device] [-x format] "
	                "[pathname...]\n",
	        myname);
	fprintf(stderr, "       %s -r -w [-ilmopuvy] [-s replstr] "
	                "[pathname...] directory\n",
	        myname);
	exit(1);
}
